### ISSUE TITLE
Apple 64-bit / M1 Feedback & Testing

### ISSUE THREAD

**Issue by leezer3** [leezer3/OpenBVE#680]
Current builds *should* (hopefully) work on 64-bit Apple / M1.

@plutommm you've tested this on M1, and it doesn't seem to work. Any chance of some more details at all?
First thing you've (hopefully obviously) got to do is to unblock it under the preferences application.

Assuming you've done this, please check that you've got Mono installed. The most recent version is 6.12.0 from:
https://www.mono-project.com/download/stable/#download-mac

The app is supposed to prompt if this isn't present, but it's really hard for me to test without an M1. 
Due to Apple / Mono, the first launch may take a minute or two, so if it's in the dock bouncing, try waiting a little longer.

cc @OmegaZeroKiller - You had a self-built copy running from the PR branch. Can you test the DMG at all?

**Comment by plutommm** [comment]
I've downloaded the newest stable version of Mono, unfortunately OpenBVE
only bounces once and stops running.
It does prompt to tell me to download the latest version of Mono when I had
version 5.0.2 on...

**Comment by leezer3** [comment]
Hmm..
I think I may have spotted a problem; Big Sur has the internal version of 11.
Let me fiddle a bit....

**Comment by leezer3** [comment]
Please try this build:
https://vps.bvecornwall.co.uk/OpenBVE/Builds/OpenBVE-2021-07-28.dmg

It should now correctly handle OS-X systems with a major version greater than 10.

**Comment by plutommm** [comment]
The app is now "running", but it's stuck on this screen...
I believe it's getting there!

**Comment by leezer3** [comment]
No picture attached.
(You might need to reply from a browser on Github for that to work)

If it's bouncing in the dock, just leave it for a while and see if it pops up eventually.

**Comment by plutommm** [comment]
Let me try sending it like this...
https://drive.google.com/file/d/1cW7X2gj-_oBZabndaRF_vanryWbLFX-w/view?usp=sharing
If you still can't see it, the screen is a very weird looking reorganized
pause screen... it only allows me to either; open route file, customise
controls, or quit game.

**Comment by leezer3** [comment]
That's absolutely as expected and as designed at the minute :)
The resolution scaling isn't perfect, and seems to be making it look a little worse on your end.

You should be able to browse the filesystem for routefiles from the Open Route File option, and this should give you a similar preview etc. to the old menu.

The other features for this are being worked on, and will be introduced sooner or later.....

TLDR:
Unfortunately, the Windows Forms toolkit used to produce the 'normal' main menu isn't available on 64-bit Mac, and so we've had to write our own menu system from scratch. This is a *much* bigger job than it sounds :)

**Comment by plutommm** [comment]
Interesting, I think it’s beginning to work as of right now, thank you for
your help.
Fighting~ :DDD

**Comment by OmegaZeroKiller** [comment]
> Current builds _should_ (hopefully) work on 64-bit Apple / M1.
> 
> @plutommm you've tested this on M1, and it doesn't seem to work. Any chance of some more details at all?
> First thing you've (hopefully obviously) got to do is to unblock it under the preferences application.
> 
> Assuming you've done this, please check that you've got Mono installed. The most recent version is 6.12.0 from:
> https://www.mono-project.com/download/stable/#download-mac
> 
> The app is supposed to prompt if this isn't present, but it's really hard for me to test without an M1.
> Due to Apple / Mono, the first launch may take a minute or two, so if it's in the dock bouncing, try waiting a little longer.
> 
> cc @OmegaZeroKiller - You had a self-built copy running from the PR branch. Can you test the DMG at all?

Sorry for the late reply, but sure thing, I can help with the testing; where can I get the DMG?

**Comment by leezer3** [comment]
Latest daily builds DMG should now be working with the GL Menu.

**Comment by plutommm** [comment]
I have just got a chance to test it
But sadly the menu is still the "not-as-preferred one" from the previous
builds.

**Comment by plutommm** [comment]
Small Question
How do I change the code setting of routes or trains to Big5/Latin-1/Shift_JIS in the new menu?

**Comment by OmegaZeroKiller** [comment]
![螢幕截圖 2021-08-15 下午5 53 12](https://user-images.githubusercontent.com/73306375/129472075-8e86fe92-a209-48c9-91f9-d76d78b7ec44.png)
Managed to get the new DMG version up and running.
However, without the settings menu, it is rather difficult to change rendering distance or preferred OpenGL version for the renderer.
Is there any way to change preferred OpenGL renderer version?

**Comment by leezer3** [comment]
Options support is on the list after I've sorted out Packages.
In the meantime, you'd have to edit the options.cfg file.
Can't remember off the top of my head exactly where this ends up on Mac, but it's likely to be in /Library/Preferences or /Library/ApplicationSupport somewhere.
Will take a look toorrow and see if I can find it.

(It'll require a bunch more menu item code so resolutions etc. can be nicely chosen, hence why I've not yet done that)

**Comment by plutommm** [comment]
Because this is my first time typing these settings in, I would like to
know how the code should look like to change the code settings to big 5.

**Comment by leezer3** [comment]
### Switching language:

Options.cfg:

```
[language]
code = zh-CN
```
This is unilkely to be the issue though. (we should be auto-detecting most charsets anyway)
I've got a partial fix on the packages branch at the minute, but I'm still working on that one; This should be solved sooner or later.

### Switching between GL1.2 and GL3:
It's all a little complicated...

The GL menu should only appear on 64-bit only / M1 Apple (e.g. Mojave and above)
Unfortunately, switching between GL1.2 and GL3 is a problem- In order to get a functioning GL3 context on Apple, we must be using the **ForwardsCompatible** flag, with a minimum requested version of **3.3** in a 64-bit process.
The context version means that switching between the two is impossible, and if we're using 64-bit, then that breaks the 'old' main menu......
(GL3 has never previously worked on Apple)

TLDR:
GL1.2 is only availble with the old menu and vice-versa.

**Comment by plutommm** [comment]
Sorry but I am actually trying to switch the route and train encodings to
Big 5, not the languages.
Also, just to be clear, does the code take effect simply by me editing the
options.cfg file?

**Comment by leezer3** [comment]
If you want to edit the route encoding, go a little further into the file:
```
[routeEncodings]
1250 = D:\program files\bve\railway\route\Plymouth- St Austell Demo\1980s\Class 104 (Stopping)- Liskeard [c].csv
```

The number at the front of the string is what you need to edit, and is the codepage number you wish to use. However, this is unlikely to make much difference, as in essentially all cases, we should automatically detect the codepage in use.

Try the daily build from today, as this should hopefully work a little better in this regard. (It switches the default font on Mac to one with a full set of unicode glyphs available)

**Comment by andylin2004** [comment]
Seems like the Aug 17 macOS build crashes when loading a train in. I was using the NYCT F line v5.0 and the Siemens R160 train to run, and it crashed when it tried to load the train in.

**Comment by leezer3** [comment]
That's odd.

Don't think I've touched trains, will investigate.

**Comment by andylin2004** [comment]
> That's odd.
> 
> Don't think I've touched trains, will investigate.

I used the July 28th build and it loaded just fine though (although it does glitch and drops the loading percentage from 52% to 50% at one point)

**Comment by leezer3** [comment]
Very odd.

This is what broke it:
https://github.com/leezer3/OpenBVE/pull/683

I can reproduce the crash with GL2, but not GL3.
What's even more odd is that it's a native crash, as opposed to something I've done. I *think* we must have hit a Mono issue somewhere, will need to try and isolate this a little further now.....

**Comment by leezer3** [comment]
I think I've found the trouble- Timing related.....

Today's build should update shortly to fix this.

TLDR:
We need a thread with a valid GL context to update the lighting. The way this was being called after the lighting event changes no longer guaranteed that; Changed so that this is now queued for update first thing in the render loop, which will guarantee this.
(No idea why this manages to take down the whole thing with a native crash)

**Comment by andylin2004** [comment]
> Build from today should now have a basic functional options menu, which will let you change screen resolution, interpolation, AA / AF and the viewing distance :)
> 
> Will likely be looking to push this in a stable build in a week or two- I think it's now at a usable state, if not pretty.

I don't think the screen resolution setting will work when the window is full screened. If I make the window full screen then try to change around the resolution setting, instead of having the window scale to fit the screen, the window will shrink such that the window pixel size is even smaller than the said window size. This is on a MacBook Air M1 with the display setting set at looks like 1680*1050.
